### PR TITLE
feat: Adds mlflow run logging

### PR DESCRIPTION
**Pull request by ElefHead**
Saw https://github.com/ultralytics/yolov5/issues/4231 and https://github.com/ultralytics/yolov5/pull/7840 and figured I'd make a PR to add some rudimentary logging for Mlflow that I've successfully used.
* Assumes that mlflow and experiment is setup by user
* Logger creates a run and tracks params, metrics, and artifacts

To make use of this code:

1. Run Mlflow ui
```bash
pip install mlflow==1.26.1
mlflow ui
```
2. Run training code. eg.
```bash
python train.py --img 640 --batch 16 --epochs 3 --data coco128.yaml --weights yolov5n.pt --cache cache --save-period 1
```

This will create a run and log all params, metrics, artifacts, and log the model to the run. If no experiment is set, everything will be logged to the default experiment. Below is an example of a run:
Screen Shot 2022-08-25 at 11 09 26 PM

Alternatively, users can also run the code as an Mlflow Project. For that users would have to create their own [MLproject](https://www.mlflow.org/docs/latest/projects.html#mlproject-file) file, specify the backend-config and conda.yaml and run the project as 

```bash
mlflow run yolov5 -p {param1} --experiment-name {experiment-name} -b {databricks or local} --backend-config {set only if running on databricks}
```
and all of this will still work.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Integration of Mlflow for experiment tracking in Ultralytics YOLOv5.

### 📊 Key Changes
- Added Mlflow (`mlflow==1.26.1`) to requirements, expanding the suite of supported experiment tracking tools.
- Introduced `MlflowLogger` class to handle Mlflow logging activities within the code.
- Enabled artifact logging (like labels, mosaics, and validation images), parameters, metrics, and model logging through Mlflow.
- Updated logger initialization code to support the new Mlflow integration.
- Created dedicated `mlflow_utils.py` module under `utils/loggers/mlflow` directory.

### 🎯 Purpose & Impact
- **Purpose**: To provide users additional flexibility to track and visualize their training experiments using Mlflow.
- **Impact**: Users can now choose Mlflow as their experiment tracking tool, potentially leveraging existing Mlflow setups and integrating with Mlflow's extensive ecosystem. This enhances the experiment management and reproducibility features of YOLOv5 training workflows. 📈🔍⏱️

### CODE REVIEW

**Comment by ElefHead** [comment]
Hi there! 

I saw https://github.com/ultralytics/yolov5/issues/4231 and https://github.com/ultralytics/yolov5/pull/7840 and figured I'd make a PR to add some rudimentary logging for Mlflow that I used for work.  

This is how a run looks.

The assumption is that mlflow and the experiment would have been setup by the user and this just creates a run and logs params, metrics, and the model as artifacts.

Here's how a sample run looks.
Screen Shot 2022-06-18 at 11 25 52 PM

Thanks!

**Comment by fcan26** [comment]
amazing work

**Comment by ElefHead** [comment]
@glenn-jocher @AyushExel  hi there! was wondering if this PR could be reviewed? TIA!

**Comment by AyushExel** [comment]
@ElefHead We currently have other higher priority tasks. We'll probably come back to this after 2 releases

**Comment by fcan26** [comment]
Hi Sir,
I think that,  we must also add the mlflow part
  In the utils/loggers/__init__.py/  (def on_pretrain_routine_end(self):) method.

**Comment by ElefHead** [comment]
>  I think that, we must also add the mlflow part In the utils/loggers/**init**.py/ (def on_pretrain_routine_end(self):) method.

@fcan26  Thanks for the review, I have added those as well as the post train batch end artifacts and the validation batch artifacts. Updated the screenshot on the description.

**Comment by jayer95** [comment]
@ElefHead 
Hi ElefHead,
I'm very interested in your integration of mlflow into yolov5, I'm also working on yolov5 and mlflow recently.

I referenced this page, but the content of this page is a bit out of date.
https://medium.com/codex/setting-up-mlflow-for-ultralytics-yolov5-1380b5f8cac5

At present, I have successfully integrated mlflow with yolov5, but the parameters and weights that can be tracked are not as rich as yours.

I set up the UI page of mlflow through the following mlflow command. I can specify the URL and Port number by myself:

mlflow server \
--backend-store-uri sqlite:///mlflow.db \
--default-artifact-root ./mlruns \
--host 0.0.0.0 --port 8000 &

A s you said, "mlflow ui" will be automatically set up at "http://127.0.0.1:5000/", how can I specify the URL and Port number by myself?

**Comment by ElefHead** [comment]
@jayer95 You can do an `export MLFLOW_TRACKING_URI=http://0.0.0.0:8000`. 

I haven't used `mlflow server` a lot because the last time I tried it out, I saw weird behavior. Maybe you can help me out.

**Comment by jayer95** [comment]
@ElefHead 
Thank you, I use "export MLFLOW_TRACKING_URI=http://0.0.0.0:8000" to get the correct mlflow tracking url.

What kind of weird behaviors are you seeing?

I found that mlflow 1.27 is not as easy to use as 1.26.1.

This is my usage steps:

`git clone https://github.com/ElefHead/yolov5.git` (yours)
`git checkout dev/mlflow-run-tracking`

`cd yolov5`

`mlflow server --backend-store-uri sqlite:///mlflow.db --default-artifact-root ./mlruns --host 0.0.0.0 --port 8000 &`

`export MLFLOW_TRACKING_URI="http://0.0.0.0:8000"`

`mlflow run . --env-manager local --experiment-name ultralytics/yolov5 -P epochs=3`

Check url: http://0.0.0.0:8000

[MLproject]
name: ultralytics/yolov5
conda_env: mlflow.yml
entry_points:
  main:
    parameters:
      img_size: {type: int, default: 640}
      batch_size: {type: int, default: 16}
      epochs: {type: int, default: 300}
      data_file: {type: string, default: "./data/coco128.yaml"}
      cfg_file: {type: string, default: "./models/yolov5s.yaml"}
      weights_file: {type: string, default: "./yolov5s.pt"}
    command: |
      python train.py \
      --img {img_size} \
      --batch {batch_size} \
      --epochs {epochs} \
      --data {data_file} \
      --cfg {cfg_file} \
      --weights {weights_file}

**Comment by ElefHead** [comment]
So, `mlflow run <local project>` is broken in `mlflow 1.27.0`, the fix will be in next release. 
Nothing major or worrying, I remember having to handle params differently - but I think the code here should all work fine. Mostly, just let me know if it doesn't work for some reason and I can try fixing it. :)

**Comment by jayer95** [comment]
@ElefHead 
I know why.
In the latest version 1.27, he asked us to step back to one level above yolov5 and execute:

`mlflow run \#yolov5 --env-manager local --experiment-name ultralytics/yolov5 -P epochs=3`

My operating system is Ubuntu, the official documents of 1.27 have instructions, I need to go to the upper directory of the project, and add "#" in front of the folder name of the project, because "#" is a symbol, so on Ubuntu, it must be add slash.

**Comment by ElefHead** [comment]
Yeah, it's because if the local folder is a git repo, it got treated as a git Uri. 

https://github.com/mlflow/mlflow/issues/6194

**Comment by jayer95** [comment]
@ElefHead 
Hi, I don't know why, I used my own data (unofficial coco128) for training and got stuck at the first epoch, as shown below,
gnome-shell-screenshot-NMUDQ1
I use "ctrl+c" to force close the training, the memory occupied by the gpu is not released, and htop can't see the pid, maybe you can help me out.

**Comment by ElefHead** [comment]
hmm, @jayer95 - unfortunately idk how to recreate your exact experiment, but I could run an mlflow server the same way you run it and try out a bunch of runs and see if I see the same problem (I'd have to do it over the weekend). Technically, the mlflow logging part should do nothing with the GPU (till the end at least where the best model is reloaded to log to mlflow) or block none of these processes, unless there's something happening with the server running on the same instance.

**Comment by jayer95** [comment]
@ElefHead 
I'll do a close experiment to try and find the problem, and at the same time, I'll prepare a reproducible dataset.
I use "mlflow run" to start training and track mlflow logs, when I use the yolov5 official coco128 dataset, everything works fine. I'm thinking if it's because of my "mlflow run" command error.

**Comment by fcan26** [comment]
> I'll do a close experiment to try and find the problem, and at the same time, I'll prepare a reproducible dataset.
> I use "mlflow run" to start training and track mlflow logs, when I use the yolov5 official coco128 dataset, everything works fine. I'm thinking if it's because of my "mlflow run" command error.

Hi!
I'm also working on the @ElefHead repo and trying to add a database. If you pull the repo directly and then run it, the project itself gives automatic uri and mlflow ui works.
If you have integrated the database, you can try this command,
mlflow ui --backend-store-uri sqlite:///mlruns.db
My database integration is not finished yet, I ran a small project with this command.

**Comment by fcan26** [comment]
+ @jayer95  Also don't forget to add this command to main part of train.py
mlflow.set_tracking_uri("sqlite:///mlruns.db")

**Comment by jayer95** [comment]
@fcan26 
Hi,
Thanks for your suggestion, about add this command to main part of train.py
`mlflow.set_tracking_uri("sqlite:///mlruns.db")`
What can it be used to do?

**Comment by fcan26** [comment]
> @fcan26 Hi, Thanks for your suggestion, about add this command to main part of train.py `mlflow.set_tracking_uri("sqlite:///mlruns.db")` What can it be used to do?
to connect database

**Comment by fcan26** [comment]
> 
when i followed your path i got this error :(
 mlflow run \#yolov5 --env-manager local --experiment-name ultralytics/yolov5 -P epochs=3
2022/08/03 08:42:25 INFO mlflow.projects: 'ultralytics/yolov5' does not exist. Creating a new experiment
2022/08/03 08:42:25 ERROR mlflow.cli: === Could not find subdirectory yolov5 of  ===

**Comment by jayer95** [comment]
@ElefHead 
I found a small bugs today,
when the weights I set during training is empty, it means that I do not use the pre-trained model, but let the model start training from scratch.

As follows:

```
mlflow run . \
--env-manager local \
--experiment-name ultralytics/yolov5 \
-P epochs=3 \
-P weights_file=''
```

When the training is complete, it will get the error:

```
Fusing layers... 
YOLOv5s summary: 213 layers, 7225885 parameters, 0 gradients, 16.4 GFLOPs
Adding AutoShape... 
Traceback (most recent call last):
  File "train.py", line 642, in <module>
    main(opt)
  File "train.py", line 537, in main
    train(opt.hyp, opt, device, callbacks)
  File "train.py", line 442, in train
    callbacks.run('on_train_end', last, best, plots, epoch, results)
  File "/home/gvsai/workspace/mlflow/yolov5/utils/callbacks.py", line 71, in run
    logger['callback'](*args, **kwargs)
  File "/home/gvsai/workspace/mlflow/yolov5/utils/loggers/__init__.py", line 219, in on_train_end
    self.mlflow.log_model(best)
  File "/home/gvsai/workspace/mlflow/yolov5/utils/loggers/mlflow/mlflow_utils.py", line 85, in log_model
    self.weights.unlink()
  File "/home/gvsai/anaconda3/envs/mlflow/lib/python3.8/pathlib.py", line 1325, in unlink
    self._accessor.unlink(self)
IsADirectoryError: [Errno 21] Is a directory: '.'
2022/08/03 17:54:20 ERROR mlflow.cli: === Run (ID '05158d98036e46c6a167c25badb23274') failed ===
```

https://github.com/ElefHead/yolov5/blob/dev/mlflow-run-tracking/utils/loggers/mlflow/mlflow_utils.py#L84

I modified it to:

`if self.weights.exists() and self.weights.is_file() and (self.weights.parent.resolve() == ROOT.resolve()):`

Problem solved, please review and refer.

**Comment by fcan26** [comment]
@jayer95
Which part did you integrate TRACKING_URI="http://0.0.0.0:8000"  into?
and did you use mlflow.client ?

**Comment by jayer95** [comment]
> > @fcan26 Hi, Thanks for your suggestion, about add this command to main part of train.py `mlflow.set_tracking_uri("sqlite:///mlruns.db")` What can it be used to do?
> > to connect database

Thank you for your explanation, I will give it a try, I have not added this line of code before. 
It seems that mlflow can automatically connect with the database through the following command:

```
mlflow server \
--backend-store-uri sqlite:///mlflow.db \
--default-artifact-root ./mlruns \
--host 0.0.0.0 --port 8000 &
```

`export MLFLOW_TRACKING_URI="http://0.0.0.0:8000"`

**Comment by fcan26** [comment]
> > > @fcan26 Hi, Thanks for your suggestion, about add this command to main part of train.py `mlflow.set_tracking_uri("sqlite:///mlruns.db")` What can it be used to do?
> > > to connect database
> 
> Thank you for your explanation, I will give it a try, I have not added this line of code before. It seems that mlflow can automatically connect with the database through the following command:
> 
> ```
> mlflow server \
> --backend-store-uri sqlite:///mlflow.db \
> --default-artifact-root ./mlruns \
> --host 0.0.0.0 --port 8000 &
> ```
> 
> `export MLFLOW_TRACKING_URI="http://0.0.0.0:8000"`

Thank u!

**Comment by jayer95** [comment]
> > 
> 
> when i followed your path i got this error :( mlflow run #yolov5 --env-manager local --experiment-name ultralytics/yolov5 -P epochs=3 2022/08/03 08:42:25 INFO mlflow.projects: 'ultralytics/yolov5' does not exist. Creating a new experiment 2022/08/03 08:42:25 ERROR mlflow.cli: === Could not find subdirectory yolov5 of ===

This seems to be a bug.
https://github.com/mlflow/mlflow/issues/6194

Which version of mlflow did you install?
If it is 1.27, please "cd .." under the folder of yolov5, go back to the previous layer, and:

```
mlflow run \#yolov5 \
--env-manager local \
--experiment-name ultralytics/yolov5 \
-P epochs=3
```

**Comment by jayer95** [comment]
> @jayer95 Which part did you integrate TRACKING_URI="http://0.0.0.0:8000" into? and did you use mlflow.client ?

I entered the following commands manually:
`export MLFLOW_TRACKING_URI="http://0.0.0.0:8000"
`

No, I didn't, I just started working on mlflow.
Thanks for discussing with me!

**Comment by fcan26** [comment]
> > @jayer95 Which part did you integrate TRACKING_URI="http://0.0.0.0:8000" into? and did you use mlflow.client ?
> 
> I entered the following commands manually: `export MLFLOW_TRACKING_URI="http://0.0.0.0:8000" `
> 
> No, I didn't, I just started working on mlflow. Thanks for discussing with me!
Its my pleasure! I also thank you. I hope we can run the project properly in local with database integration.

**Comment by ElefHead** [comment]
@jayer95 
> I found a small bugs today, ...

Thanks! I can update it

**Comment by jayer95** [comment]
@ElefHead 
I reset the server with the following command:

```
mlflow server \
--backend-store-uri sqlite:///mlflow.db \
--artifacts-destination s3://mlflowbucket \
--serve-artifacts \
--host 0.0.0.0 \
--port 8000
```

Now I can upload the "mlruns" to aws s3.

**Comment by fcan26** [comment]
@jayer95 My terminal state like this  >>>  ~/Desktop/x/yolov5$
I run this commet   >>   mlflow run \#yolov5 \ --env-manager local \ --experiment-name ultralytics/yolov5 \ -P epochs=3
but I got this error:   >>     2022/08/05 04:31:40 INFO mlflow.projects: 'ultralytics/yolov5' does not exist. Creating a new experiment
2022/08/05 04:31:40 ERROR mlflow.cli: === Could not find subdirectory yolov5 of  ===


I run this repo locally(without aws I mean), also I follow this article (https://medium.com/codex/setting-up-mlflow-for-ultralytics-yolov5-1380b5f8cac5 ), but I didnt use docker.
Here are the steps I followed
pull the @ElefHead 's repo
create MLproject
> mlflow server --backend-store-uri sqlite:///mlflow.db --default-artifact-root ./mlruns --host 0.0.0.0 --port 8000 &
> export MLFLOW_TRACKING_URI="http://0.0.0.0:8000"
My terminal state like this  >>>  ~/Desktop/x/yolov5$
I run this commet   >>   mlflow run \#yolov5 \ --env-manager local \ --experiment-name ultralytics/yolov5 \ -P epochs=3
**but I got this error**: >     2022/08/05 04:31:40 INFO mlflow.projects: 'ultralytics/yolov5' does not exist. Creating a new experiment
2022/08/05 04:31:40 ERROR mlflow.cli: === Could not find subdirectory yolov5 of  ===
I will be glad if you can help..

**Comment by fcan26** [comment]
> 

Now my error like this:

~/Desktop/x/yolov5$ mlflow run \#yolov5 \
> --env-manager local \
> --experiment-name ultralytics/yolov5 \
> -P epochs=3
2022/08/05 05:39:40 ERROR mlflow.cli: === Could not find subdirectory yolov5 of  ===

**Comment by fcan26** [comment]
> > 
> 
> Now my error like this:
> 
> ~/Desktop/x/yolov5$ mlflow run #yolov5 \
> 
> > --env-manager local 
> > --experiment-name ultralytics/yolov5 
> > -P epochs=3
> > 2022/08/05 05:39:40 ERROR mlflow.cli: === Could not find subdirectory yolov5 of  ===


@ElefHead @jayer95 
I create mlflow.yml like this

name: mlflow_example
channels:
- defaults
- conda-forge
- pytorch
dependencies:
  - python=3.8.8
  - pip
  - pip:
    - torch>=1.7.0
    - torchvision>=0.8.1
    - mlflow=1.26.1  
    - tensorboardX
    - protobuf<=3.20.1

**Comment by ElefHead** [comment]
@jayer95 That's great :) 
Does everything work as expected?

**Comment by jayer95** [comment]
@ElefHead 
Currently, mlflow branch is working well.
I found that the problem mentioned earlier about GPU memory leaks may be related to my 3090 under high load for a long time. When I restarted the computer, the problem was solved.
@fcan26 
And you?

**Comment by fcan26** [comment]
Since 'clearml' added the part, the repo doesn't work for me.
I run with this comment
python train.py --img 640 --batch 16 --epochs 3 --data coco128.yaml --weights yolov5s.pt --cache

I got this error
YOLOv5 temporarily requires wandb version 0.12.10 or below. Some features may not work as expected.
Traceback (most recent call last):
  File "train.py", line 634, in <module>
    main(opt)
  File "train.py", line 529, in main
    train(opt.hyp, opt, device, callbacks)
  File "train.py", line 92, in train
    loggers = Loggers(save_dir, weights, opt, hyp, LOGGER)  # loggers instance
  File "/Desktop/fcan/yolov5-dev-mlflow-run-tracking/utils/loggers/__init__.py", line 110, in __init__
    self.clearml = ClearmlLogger(self.opt, self.hyp)
  File "/Desktop/fcan/yolov5-dev-mlflow-run-tracking/utils/loggers/clearml/clearml_utils.py", line 84, in __init__
    self.task = Task.init(
  File "/.local/lib/python3.8/site-packages/clearml/task.py", line 581, in init
    task = cls._create_dev_task(
  File "/.local/lib/python3.8/site-packages/clearml/task.py", line 2998, in _create_dev_task
    task = cls(
  File "/.local/lib/python3.8/site-packages/clearml/task.py", line 182, in __init__
    super(Task, self).__init__(**kwargs)
  File "/.local/lib/python3.8/site-packages/clearml/backend_interface/task/task.py", line 155, in __init__
    super(Task, self).__init__(id=task_id, session=session, log=log)
  File "/.local/lib/python3.8/site-packages/clearml/backend_interface/base.py", line 145, in __init__
    super(IdObjectBase, self).__init__(session, log, **kwargs)
  File "/.local/lib/python3.8/site-packages/clearml/backend_interface/base.py", line 39, in __init__
    self._session = session or self._get_default_session()
  File "/.local/lib/python3.8/site-packages/clearml/backend_interface/base.py", line 115, in _get_default_session
    InterfaceBase._default_session = Session(
  File "/.local/lib/python3.8/site-packages/clearml/backend_api/session/session.py", line 178, in __init__
    raise ValueError(
ValueError: ClearML configuration could not be found (missing `~/clearml.conf` or Environment CLEARML_API_HOST)
To get started with ClearML: setup your own `clearml-server`, or create a free account at https://app.clear.ml
wandb: Waiting for W&B process to finish... (failed 1). Press Control-C to abort syncing.


The mlruns folder is not created and the wandb sync is broken.

It was working fine before adding clearml, I'm still working with the old repo.

@ElefHead @jayer95

**Comment by ElefHead** [comment]
@fcan26 could you check if you're using commit [2fc333a](https://github.com/ultralytics/yolov5/pull/8261/commits/2fc333af585f07d4c7e9abef77b72bc29b69419c) or later?

Also, I think you might want to follow the tutorial in tutorials.ipynb notebook for using clearml. If you don't want to use it, you can avoid installing it or if you have installed the python package, you can uninstall it.

**Comment by fcan26** [comment]
@ElefHead Problem solved, thank you for your advice!

**Comment by alrightkami** [comment]
Hey! Thank you for the repo, it is very useful! I've modified it a bit for my case and noticed that as soon as I connect S3, I get the following exception:

```
Traceback (most recent call last):
  File "/home/jovyan/data/kamila/yolov5_mlflow/yolov5/train.py", line 636, in <module>
    main(opt)
  File "/home/jovyan/data/kamila/yolov5_mlflow/yolov5/train.py", line 529, in main
    train(opt.hyp, opt, device, callbacks)
  File "/home/jovyan/data/kamila/yolov5_mlflow/yolov5/train.py", line 337, in train
    callbacks.run('on_train_batch_end', ni, model, imgs, targets, paths, plots)
  File "/home/jovyan/data/kamila/yolov5_mlflow/yolov5/utils/callbacks.py", line 71, in run
    logger['callback'](*args, **kwargs)
  File "/home/jovyan/data/kamila/yolov5_mlflow/yolov5/utils/loggers/__init__.py", line 153, in on_train_batch_end
    [self.mlflow.log_artifacts(f, "train") for f in files if f.exists()]
  File "/home/jovyan/data/kamila/yolov5_mlflow/yolov5/utils/loggers/__init__.py", line 153, in <listcomp>
    [self.mlflow.log_artifacts(f, "train") for f in files if f.exists()]
  File "/home/jovyan/data/kamila/yolov5_mlflow/yolov5/utils/loggers/mlflow/mlflow_utils.py", line 87, in log_artifacts
    self.mlflow.log_artifact(artifact.resolve(), artifact_path=relpath)  # custom_mlflow
  File "/opt/conda/lib/python3.9/site-packages/mlflow/tracking/fluent.py", line 752, in log_artifact
    MlflowClient().log_artifact(run_id, local_path, artifact_path)
  File "/opt/conda/lib/python3.9/site-packages/mlflow/tracking/client.py", line 955, in log_artifact
    self._tracking_client.log_artifact(run_id, local_path, artifact_path)
  File "/opt/conda/lib/python3.9/site-packages/mlflow/tracking/_tracking_service/client.py", line 365, in log_artifact
    artifact_repo.log_artifact(local_path, artifact_path)
  File "/opt/conda/lib/python3.9/site-packages/mlflow/store/artifact/s3_artifact_repo.py", line 124, in log_artifact
    self._upload_file(
  File "/opt/conda/lib/python3.9/site-packages/mlflow/store/artifact/s3_artifact_repo.py", line 117, in _upload_file
    s3_client.upload_file(Filename=local_file, Bucket=bucket, Key=key, ExtraArgs=extra_args)
  File "/opt/conda/lib/python3.9/site-packages/boto3/s3/inject.py", line 143, in upload_file
    return transfer.upload_file(
  File "/opt/conda/lib/python3.9/site-packages/boto3/s3/transfer.py", line 281, in upload_file
    raise ValueError('Filename must be a string')
ValueError: Filename must be a string
```

But it's very easily fixed if you change line 87 in mlflow_utils to this:
`self.mlflow.log_artifact(str(artifact.resolve()), artifact_path=relpath)`

**Comment by ElefHead** [comment]
Thanks @alrightkami i'll fix it

**Comment by alrightkami** [comment]
Also, I've noticed that the whole yolov5 folder (with train.py & detect.py, all .yaml's, etc.) is being logged as well which I find a bit redundant to store for every run. 
Where can I disable it?

image

**Review by dankernel — COMMENTED** [review]
It works well with mlflow in my environment.

**Comment by ElefHead** [comment]
@dankernel thanks for reviewing :)

**Review by dankernel — COMMENTED** [review]
It works well with mlflow in my environment.
I hope it will be merged into main branch.

**Comment by ElefHead** [comment]
@AyushExel @glenn-jocher Hi, I have been keeping this fork current so that whenever you're free to review, it can be done quickly. Just wanted to point out that for some reason the CI test for macos latest 3.10 is failing and was hoping you could point out why. I have a macOS machine and am able to test with python3.10.6 and it works for me.

**Comment by glenn-jocher** [comment]
@ElefHead hi thanks! Yes macos-latest CI seems to fail occasionally on yolov5n-cls.pt download. I don't know why.

**Comment by ahmadmustafaanis** [comment]
please integrate this. thanks

**Comment by macfly1202** [comment]
Hello, is there any chance that this merge request will be included in the master  branch? As in compagny, on-premise execution, mlflow is the only app we can use without cloud provider storage (wandb, comet, ...)

**Comment by mmortazavi** [comment]
A must feature that has been missing in this awesome YoloV5 implementation! Hope the heavy load of work that is done is going to be merged soon in he master branch and we have a flawless access to mlflow tracking! Thanks

**Comment by sebos123** [comment]
@dankernel any update here?

**Comment by jh-park-joyfulturtles** [comment]
@sebos123 This is already enough.

**Comment by dankernel** [comment]
@sebos123 This is already enough.

**Comment by ElefHead** [comment]
> @ElefHead We currently have other higher priority tasks. We'll probably come back to this after 2 releases

@glenn-jocher @AyushExel Hi, happy new year! I was wondering if this PR can be reviewed now?

**Comment by furkannnnnnnnn** [comment]
Encountered the following error on yolov5 `version v7.0` during segmentation training, with mlflow version: `2.1.1`:

```
Traceback (most recent call last):
  File "segment/train.py", line 658, in <module>
    main(opt)
  File "segment/train.py", line 554, in main
    train(opt.hyp, opt, device, callbacks)
  File "segment/train.py", line 382, in train
    logger.log_metrics(metrics_dict, epoch)
  File "/home/ubuntu/yolov5/utils/loggers/__init__.py", line 401, in log_metrics
    self.mlflow.log_metrics(metrics, epoch=epoch)
  File "/home/ubuntu/yolov5/utils/loggers/mlflow/mlflow_utils.py", line 129, in log_metrics
    self.mlflow.log_metrics(metrics=metrics_dict, step=epoch)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/mlflow/tracking/fluent.py", line 674, in log_metrics
    MlflowClient().log_batch(run_id=run_id, metrics=metrics_arr, params=[], tags=[])
  File "/home/ubuntu/.local/lib/python3.8/site-packages/mlflow/tracking/client.py", line 965, in log_batch
    self._tracking_client.log_batch(run_id, metrics, params, tags)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/mlflow/tracking/_tracking_service/client.py", line 394, in log_batch
    self.store.log_batch(run_id=run_id, metrics=metrics_batch, params=[], tags=[])
  File "/home/ubuntu/.local/lib/python3.8/site-packages/mlflow/store/tracking/file_store.py", line 1014, in log_batch
    _validate_batch_log_data(metrics, params, tags)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/mlflow/utils/validation.py", line 295, in _validate_batch_log_data
    _validate_metric(metric.key, metric.value, metric.timestamp, metric.step)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/mlflow/utils/validation.py", line 120, in _validate_metric
    _validate_metric_name(key)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/mlflow/utils/validation.py", line 95, in _validate_metric_name
    raise MlflowException(
mlflow.exceptions.MlflowException: Invalid metric name: 'metrics/precision(B)'. Names may only contain alphanumerics, underscores (_), dashes (-), periods (.), spaces ( ), and slashes (/).
```

---

Update:

https://github.com/ultralytics/yolov5/blob/064365d8683fd002e9ad789c1e91fa3d021b44f0/utils/segment/metrics.py#L180-L210

Changing `(B)` to `_B` and `(M)` to `_M` on relevant metrics solves the issue.

**Comment by sasikumari2i** [comment]
Hi @dankernel, Is there any update on merging this PR  ?

**Comment by EricKeenan** [comment]
@ElefHead Thanks for posting this PR. We found this useful! 

Curious if you have tackled MLflow integration for resuming paused/failed runs? Haven't been able to make MLflow integration using
```bash
python -m torch.distributed.launch --nproc_per_node 4 train.py --resume
```

It does work if you start from scratch though.

**Comment by glenn-jocher** [comment]
@momocici Thank you! I am glad it has been useful for you.

Regarding resuming runs, the `--resume` flag should work fine when resuming the run. MLflow will automatically attach to the existing run ID if run with the same name. In the case of a failed run, the tracking info should be still available, so it should be possible to resume from there.

If you are still encountering issues, you may want to check if the same experiment and run name are being used or if there are any inconsistencies between the saved models and logged metrics.

File: utils/loggers/mlflow/mlflow_utils.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        only to log initial training parameters.
+
+        Args:
+            params (Dict[str, Any]): Parameters as dict
+        """
+        try:
+            flattened_params = MlflowLogger._flatten_params(params_dict=params)
+            run = self.client.get_run(run_id=self.run_id)
+            logged_params = run.data.params
+            [
```
Comment (YoRigid) [inline utils/loggers/mlflow/mlflow_utils.py:110]: May 26th, 2023, the first iteration in the list comprehension fails, triggering the generic try and except, and skipping every param, perhaps change the list comprehension to a for loop, have the try and except inside the for loop?

```python
        newly_logged = []
        for key, value in flattened_params.items():
            if (
                key in logged_params or not value
            ):  # TODO find out if mlflow.log_param 'Appends' or 'Overwrites'
                continue
            try:
                str(value).strip()
                assert value
                newly_logged.append(mlflow.log_param(key, str(value)))
            except AssertionError:
                LOGGER.info("%s, not logging None value: %s", LOG_PREFIX, key)

**Comment by vfc2** [comment]
@glenn-jocher Hi there, I have made a similar implementation of mlflow logging targeted towards Azure ML (their Job GUI is populated by mlflow as it became the defacto logging framework of their Python SDK v2). I was about to PR my changes and I found this… I’m not sure why it was never merged, is it even worth the effort of creating a new PR?

Example of Azure ML mlflow logging integration:
image

**Comment by glenn-jocher** [comment]
@vincent-f2 hi there! It's great to hear about your implementation for Azure ML logging with MLflow. The integration of MLflow or any other logging framework can indeed be quite valuable for many users.

Regarding the previous PR, it's not uncommon for PRs to take some time to merge, especially if they introduce new features or integrations that require thorough testing and review to ensure they align with the project's goals and maintainability.

If you believe your implementation could benefit the community and you're willing to align it with the current state of the repository, I encourage you to submit a PR. We always appreciate contributions that can enhance the functionality of YOLOv5. Just make sure to provide a clear description of the changes, how they integrate with Azure ML, and any necessary documentation or tests.

Remember, even if there's a delay in reviewing or merging, your contribution could spark discussions that lead to improvements or alternative solutions. Your effort is valuable, and we look forward to seeing your PR! 🚀

**Pull request closed without merging**